### PR TITLE
Add hash of source files in debug info

### LINKED ISSUE #68980: Consider emitting LLVM DIFile checksums.
We already hash the contents of `SourceFile`s to determine whether the file on disk (if it even exists) still has the same contents, across crates.

I came across [LLVM supporting MD5 and SHA1 checksums](https://llvm.org/docs/LangRef.html#difile) for `DIFile`s, and I'm wondering if it could be useful in a similar way - perhaps debuggers would be able to detect changed sources?

cc @michaelwoerister

### PR DESCRIPTION
**Pull request by arlosi**
LLVM supports placing the hash of source files inside the debug info.
This information can be used by a debugger to verify that the source code matches
the executable.

This change adds support for both hash algorithms supported by LLVM, MD5 and SHA1, controlled by a target option.

* DWARF only supports MD5
* LLVM IR supports MD5 and SHA1 (and SHA256 in LLVM 11).
* CodeView (.PDB) supports MD5, SHA1, and SHA256.

Fixes #68980.

Tracking issue: #70401

rustc dev guide PR with further details: https://github.com/rust-lang/rustc-dev-guide/pull/623

### CODE REVIEW

File: src/librustc_session/options.rs
Comment (bjorn3) [inline src/librustc_session/options.rs]: Why not always embed a `SHA-1` checksum?

Comment (arlosi) [reply src/librustc_session/options.rs]: Ultimately, I'd like to add `SHA-256` and make that the default, however, LLVM does not support it yet. I do have a patch to add `SHA-256` to LLVM, but I haven't sent it to them yet.

If we made `SHA-1`, the default now, would we be comfortable changing the default in the future to `SHA-256`?

Comment (bjorn3) [reply src/librustc_session/options.rs]: I don't think we provide any stability guarantees regarding the exact DWARF being emitted. I think it only matters that the emitted DWARF makes a program debug-able using gdb, lldb and maybe other debuggers. Changing the hash algorithm shouldn't have any effect on that.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs
Comment (rylev) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: This line isn't necessary and can be removed.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs
Comment (rylev) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: You can replace the second value of the tuple with `String::new` (which doesn't allocate):
```rust
let (checksum_kind, checksum_value) = match checksum {
  Some(checksum) => {
      let kind = match checksum.kind {
          rustc_span::SourceFileChecksumKind::Md5 => llvm::ChecksumKind::MD5,
          rustc_span::SourceFileChecksumKind::Sha1 => llvm::ChecksumKind::SHA1,
       };
  
      (kind, hex::encode(checksum.bytes()))
   }
   None => (llvm::ChecksumKind::None, String::new()),
}
```

File: src/librustc_span/lib.rs
Comment (rylev) [inline src/librustc_span/lib.rs]: Nit: This should probably `as_bytes` which is the standard borrow to borrow method naming convention. [See String as an example.](https://doc.rust-lang.org/std/string/struct.String.html#method.as_bytes)

File: src/librustc_span/lib.rs
Comment (rylev) [inline src/librustc_span/lib.rs]: Would it make more sense to use `?` instead of `and_then`?

Comment (arlosi) [reply src/librustc_span/lib.rs]: Maybe you meant `map`? I don't think I can use the `?` operator here.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs
Comment (arlosi) [reply src/librustc_codegen_llvm/debuginfo/metadata.rs]: Thanks. That's much nicer.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs
Comment (bjorn3) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Maybe handroll a hex encoder? That would save a dependency.

File: src/librustc_session/options.rs
Comment (bjorn3) [inline src/librustc_session/options.rs]: This affects the output object files, so it should be `TRACKED`. You could also save the value of the previous compilation somewhere and then make the [codegen dep node](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/mono/struct.CodegenUnit.html#method.codegen_dep_node) dirty [here](https://github.com/rust-lang/rust/blob/57e1da59cd0761330b4ea8d47b16340a78eeafa9/src/librustc_codegen_llvm/base.rs#L108) if it changes.

**Comment by eddyb** [comment]
Seems like this would fix #68980.

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: See `pub src_hash: u128,` - we already have a hash, it's just not MD5 or SHA-1 (it's probably Blake?). We should maybe consolidate the two so we're not computing several hashes of the same file.

cc @michaelwoerister @petrochenkov

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: We should probably do the same with `src_hash` then? Ideally they get computed all at once.

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: Would be nice to use a fixed-size array here, or a pair of `u128` and `u32` (for SHA-1).

File: src/librustc_session/options.rs
Comment (eddyb) [inline src/librustc_session/options.rs]: What are the chances we can avoid supporting MD5 altogether? What *needs* it?

Comment (eddyb) [reply src/librustc_session/options.rs]: Nevermind, already answered in https://github.com/rust-lang/rust/pull/69718#discussion_r388604692.

**Comment by nikomatsakis** [comment]
So I opened a quick [Zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/proposal.3A.20add.20cryptographic.20hashes.20to.20debuginfo.20.2369718) to discuss. I'd like to see this move forward. One thing I would like to see before we land it though is

* a quick write-up of -- nothing too long -- of what this feature does. For now this can just be a gist, we'll figure out where to save it later. I think it should include 
    * [ ] one or two paragraph english description of what the feature does 
    * [ ] some motivation for why it's useful is always good
    * [ ] some links to LLVM docs or DWARF debuginfo docs about the feature
    * [ ] reference to whether there exist equivalent flags on C++ compilers
    * [ ] any particular impl concerns or downsides we should be aware of
* we would create a tracking issue that references this write-up
* documentation for the feature in the [unstable book], the sources for which are in the repo. You could add a section to [the compiler flags](https://github.com/rust-lang/rust/tree/master/src/doc/unstable-book/src/compiler-flags) section as part of the PR. This should probably be a subset of the write-up
* side note that it'd be nice to have a rustc-dev-guide chapter (looks like it wouldn't have to be long) about this, probably just the text of the write-up, we can worry about that later

File: src/librustc_session/options.rs
Comment (eddyb) [reply src/librustc_session/options.rs]: Can you put this information in the PR description? Maybe even in the command, e.g.:
```
either `md5` (DWARF/PDB) or `sha1` (PDB-only)
```

The hash we already use (`src_hash`) is (or we truncate it to) 128-bit, so using MD5 wouldn't be much worse (except for it being vulnerable to certain attacks, which I don't know how much we care about).

**Comment by eddyb** [comment]
@arlosi @rylev How would you feel about keeping this MD5-only to start, replacing the existing `src_hash` in `SourceFile` instead of having two hashes, and having the flag *only* toggle whether LLVM gets the hash and puts it into DWARF/PDF output?

And we could even turn that flag by default (or not have a flag at all), that part I don't have too many opinions about.

File: src/librustc_span/lib.rs
Comment (michaelwoerister) [reply src/librustc_span/lib.rs]: `src_hash` is the 128 bit version of SipHash. It would not consider that cryptographically secure (although both MD5 and SHA-1 are not much better at this point, right?)

Comment (eddyb) [reply src/librustc_span/lib.rs]: Oh, we're not using BLAKE2? (anymore?) Then switching to md5 seems fine with me.

Comment (michaelwoerister) [reply src/librustc_span/lib.rs]: No, we moved away from BLAKE2 because it was unnecessarily slow and we don't need the strong guarantees it gives: https://github.com/rust-lang/rust/issues/41215

**Comment by arlosi** [comment]
@eddyb storing the hash algorithm in the target spec makes sense to me. I've updated the PR to reflect this change.

Because the `SourceMap` needs to be constructed before the target is known, I had to make the checksum kind mutable, and update it after the target is parsed. Let me know if you see a better way to do this.

**Comment by eddyb** [comment]
@arlosi Can you make creating a `SourceFile` fail before that happens?
Alternatively, `SourceMap` construction can potentially be delayed, I recall looking into something like this a long time ago.

File: Cargo.lock
Comment (bjorn3) [inline Cargo.lock]: `pest_meta` already depends on a version of `sha-1` (0.7.0). Could you please depend on the same version here to avoid duplicating `sha-1`, `digest` and `generic-array`? Also `md-5` depends on the newer `digest`. Another option would be to update `pest_meta` to `2.1.3`, which depends on `sha-1` version `0.8` instead.

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: This should derive `Copy` and not be treated different than other `Copy` types (methods on it taking `&self` is fine, however). This will signal the intent that this doesn't use some heap allocation.

Comment (eddyb) [inline src/librustc_span/lib.rs]: I wonder if you would be okay with using `Hash` instead of `Checksum` throughout here.
It's pretty arbitrary, just that `Hash` is the common name in Rust, IME.

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: This should go back to `8` (see also below in `Decodable`).

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: Since the only call of this seems to call `.as_bytes()` on a string, this should take `&str`.

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: Also this could take `&str`, I suppose.

File: src/librustc/ich/impls_syntax.rs
Comment (eddyb) [inline src/librustc/ich/impls_syntax.rs]: `src_hash` itself should implement `HashStable` (you can even try to add `#[derive(HashStable_Generic)]` on the definition, I forget which cases it works in).

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: I don't think `as_mut_bytes` is needed, as there's effectively only one call to it, identical across all 3 match arms here.

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: I wonder if `hash_bytes` is a better method name. Similarly, `hash_len` instead of `len`.
Maybe redundant?

File: src/librustc_span/lib.rs
Comment (eddyb) [inline src/librustc_span/lib.rs]: If we do this, I suppose we can just remove `Stable` eventually, and let `Md5` be the default.

File: src/libserialize/serialize.rs
Comment (eddyb) [inline src/libserialize/serialize.rs]: This is compiler-internal so you can just use const generics, i.e.:
```rust
impl<T: Encodable, const N: usize> Encodable for [T; N] {
```
I doubt the body needs to change.

Comment (eddyb) [reply src/libserialize/serialize.rs]: On second thought, `Decodable` won't work like this for now, so you can leave these changes alone.

File: Cargo.lock
Comment (arlosi) [reply Cargo.lock]: I went with upgrading `pest_meta`, which seemed to clean up the lock file.

File: src/librustc_span/lib.rs
Comment (arlosi) [reply src/librustc_span/lib.rs]: Sure, that seems fine to me. I left it as Checksum in the LLVM ffi, since that's what they call it.

File: src/librustc/ich/impls_syntax.rs
Comment (arlosi) [reply src/librustc/ich/impls_syntax.rs]: HashStable_Generic seemed to work here. Thanks!

File: src/librustc_session/session.rs
Comment (arlosi) [inline src/librustc_session/session.rs]: @eddyb I was able to delay the source map construction until here. It only needs to make it a few more lines.

However there is a circular dependency:
- constructing a SourceMap requires knowing the hash algorithm
- the hash algorithm comes from the Target specification
- reading the Target can fail and produce a Diagnostic
- producing a Diagnostic requires a SourceMap

Is there a good way we can raise a diagnostic without a SourceMap when reading the target information?

**Comment by nikomatsakis** [comment]
Nominating to give warning in the @rust-lang/compiler -- we plan on adding cryptographic hashes into the debuginfo. The format of the hash will (for now) vary depending per platform, and there will be a `-Z` flag to customize it. We can decide whether to add a stable way of customizing the hash format in the future (some options discussed were to put it in the target options or add a `-C` flag). 

@arlosi did a [writeup](https://gist.github.com/arlosi/27162cfbd4bfc9bfabc25bfbe4777160) here that describes briefly the situation with other compilers and so forth. (Thanks!)

cc @spastorino can you add this issue to the agenda for the upcoming triage meeting?

**Comment by rylev** [comment]
Does the new unstable `-Z` option still need to be briefly documented in the [unstable book](https://doc.rust-lang.org/nightly/unstable-book)?

File: src/librustc_session/session.rs
Comment (arlosi) [reply src/librustc_session/session.rs]: I resolved this by using an `early_error` for failures parsing the target information.

File: src/librustc_session/options.rs
Comment (est31) [inline src/librustc_session/options.rs]: You use the c word here.

**Comment by est31** [comment]
I'd like to interject for a moment. What you are referring to as cryptographic hash functions are actually formerly cryptographic hash functions. They should not be used for cryptographic purposes so it's wrong to call them cryptographic in 2020. In this place they don't seem to fulfill an actual security purpose any way, only to help avoid confusion when debugging software. Outside of the PR title and discussion in the PR, i've seen the word "cryptographic" only used once in the PR. I've highlighted it above. Would be nice if its uses could be removed.

File: src/librustc_session/options.rs
Comment (arlosi) [reply src/librustc_session/options.rs]: Agreed. 'Cryptographic' isn't necessary here.

Comment (est31) [reply src/librustc_session/options.rs]: Thanks again! :heart:

**Comment by nikomatsakis** [comment]
We discussed this in [today's compiler team meeting](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/weekly.20meeting.202020-03-26.20.2354818/near/191895317) and announced intent to merge. No objections raised.

File: src/librustc_span/source_map.rs
Comment (eddyb) [inline src/librustc_span/source_map.rs]: Could use a comment, but the main thing I want to say is that I'm glad this no longer needs to be mutable!

**Review by eddyb — APPROVED** [review]
LGTM, r=me if this isn't waiting on anything else.

**Comment by spastorino** [comment]
Removing nomination, this was already discussed and it's about to be merged :).

**Comment by arlosi** [comment]
@eddyb could you re-approve? I resolved merge conflicts.

**Pull request merged**
